### PR TITLE
SALTO-3173: Added support for fetching instances within nested folders.

### PR DESCRIPTION
**Pull request by tamtamirr**
Previous to this PR, we only supported fetching instances and folders that are in the first level hierarchy. This PR adds support of fetching instances within nested folders, by providing the full folder path in the include config for the folder type.

---

No noise suppression is needed here.

---

_Release Notes_: 
Salesforce Adapter:
- Support fetching of instances within nested folders.

---
_User Notifications_: 
_None_

### CODE REVIEW

**Comment by tamtamirr** [comment]
@adamro @amitgoldarb 

Still need to add UTs, and refine the code. This is a possible solution that works and fetches nested instances when a full instance name is provided through the config (similar to what SFDX do). You can see a fetched nested instance [here](https://github.com/salto-io/tamir-sf/blob/SALTO-3173/salesforce/Records/Report/NestedFolder_Copy_of_TestReport_R9h.nacl) and a fetched nested folder instance [here](https://github.com/salto-io/tamir-sf/blob/SALTO-3173/salesforce/Records/ReportFolder/NestedFolder.nacl).

**Comment by tamtamirr** [comment]
NOTE since this solution wont suffice our customer needs, no need to go over this PR yet. Currently working on a solution that fetches all the records within a given folder path.

File: packages/salesforce-adapter/src/adapter.ts [hunk 1]
```diff
@@ -555,7 +555,7 @@ export default class SalesforceAdapter implements AdapterOperations {
   Promise<FetchElements<InstanceElement[]>> {
     const typeName = await apiName(type)
     const { elements: fileProps, configChanges } = await listMetadataObjects(
-      this.client, typeName, [],
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/adapter.ts:558]: Why this change? This changes the behaviour for all of the types not just the one's we are trying to edit.

File: packages/salesforce-adapter/src/types.ts [hunk 2]
```diff
@@ -58,6 +58,7 @@ export type MetadataInstance = {
   metadataType: string
   namespace: string
   name: string
+  fileName: string
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/types.ts:61]: Should this be optional? Is this backwards compatibale?

File: packages/salesforce-adapter/src/fetch.ts [hunk 3]
```diff
@@ -77,24 +78,60 @@ export const fetchMetadataType = async (
...
+      ...props,
+      fileName: props.fileName.replace(folderName, fullPath),
+    }
+    : props
+}
+
 export const listMetadataObjects = async (
   client: SalesforceClient,
   metadataTypeName: string,
-  folders: string[],
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:83]: Do we want to change the behaviour for all metadata types? Should we make this optional?

File: packages/salesforce-adapter/src/fetch.ts [hunk 4]
```diff
@@ -225,39 +264,21 @@ export const retrieveMetadataInstances = async ({
...
       // We get folders as part of getting the records inside them
       return []
     }
-    const folders = await getFolders(type)
-    const folderNames = folders.map(folder => (folder === undefined ? folder : folder.fullName))
-    const { elements: res, configChanges: listObjectsConfigChanges } = await listMetadataObjects(
-      client, await apiName(type), folderNames.filter(isDefined)
-    )
+    const typeName = await apiName(type)
+    const { elements: res, configChanges: listObjectsConfigChanges } = isInFolderMetadataType(typeName)
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:277]: If I understand the previous code, we used to run this for every metadata type, now we are doing this for only a handful of types. This might create regressions in types we are unaware of no?

File: packages/salesforce-adapter/src/fetch.ts [hunk 5]
```diff
@@ -77,24 +78,60 @@ export const fetchMetadataType = async (
...
   }
 }
 
+const listMetadataObjectsWithinFolders = async (
+  client: SalesforceClient,
+  metadataTypeName: InFolderMetadataType,
+  folderPathByName: Record<string, string>,
+  isUnhandledError?: ErrorFilter,
+): Promise<FetchElements<FileProperties[]>> => {
+  const folders = await listMetadataObjects(client, METADATA_TYPE_TO_FOLDER_TYPE[metadataTypeName])
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:114]: I'm not sure we can be so hard-coded about the types we want to support in this way, do we have to be?

File: packages/salesforce-adapter/src/fetch.ts [hunk 6]
```diff
@@ -77,24 +78,60 @@ export const fetchMetadataType = async (
...
+  )
+  return {
+    elements: [
+      ...result.map(props => withFullPath(props, folderPathByName)),
+      ...folders.elements.map(props => withFullPath(props, folderPathByName)),
+    ],
+    configChanges: [
+      ...(errors ?? []).map(createListMetadataObjectsConfigChange),
+      ...folders.configChanges,
+    ],
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:131]: Why is this so different than previously?

File: packages/salesforce-adapter/src/fetch.ts [hunk 7]
```diff
@@ -165,12 +202,14 @@ export const fetchMetadataInstances = async ({
       prop => ({
         name: getFullName(prop),
         namespace: getNamespace(prop),
+        fileName: prop.fileName,
       })
     ).filter(
-      ({ name, namespace }) => metadataQuery.isInstanceMatch({
+      ({ name, namespace, fileName }) => metadataQuery.isInstanceMatch({
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:208]: I'm slightly confused, why did this change?

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 8]
```diff
@@ -13,15 +13,24 @@
...
+} from './metadata_types'
+
+const { isDefined } = values
+
 
 export type MetadataQuery = {
   isTypeMatch: (type: string) => boolean
   isInstanceMatch: (instance: MetadataInstance) => boolean
   isPartialFetch: () => boolean
+  getFoldersByName: (metadataType: InFolderMetadataType) => Record<string, string>
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:33]: Why is a folder a Record<string, string>?
I don't understand this interface. Maybe comment with an example?

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 9]
```diff
@@ -42,11 +51,18 @@ const DEFAULT_NAMESPACE_MATCH_ALL_TYPE_LIST = [
   'InstalledPackage',
 ]
 
+
 const getDefaultNamespace = (metadataType: string): string =>
   (DEFAULT_NAMESPACE_MATCH_ALL_TYPE_LIST.includes(metadataType)
     ? '.*'
     : DEFAULT_NAMESPACE)
 
+const getAllowedStringsFromRegex = (regexString: string): string[] => (
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:60]: Isn't this the entire point of a regex? This name doesn't add much information IMO.

**Review by amitgoldarb — COMMENTED** [review]
Added some comments.
This is a pretty basic and fundamental part of our adapter and not sure I understand it very well. Tried to be pretty defensive in the PR, hence lot's of question.

File: packages/salesforce-adapter/src/adapter.ts (hunk 1 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/adapter.ts:558]: `listMetadataObjects` used to handle two scenarios: One for `InFolderMetadataType` and one for the rest.
When I've noticed this diversion, I thought it's a huge code smell that we actually have two branches within this method.  We can discuss this F2F though :)

File: packages/salesforce-adapter/src/types.ts (hunk 2 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/types.ts:61]: This is data that always existed, but was never picked. Shouldn't be optional

File: packages/salesforce-adapter/src/fetch.ts (hunk 3 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts:83]: The rest of the metadata types always passed empty array here. This method handled two scenarios and I've decided to split them when the code became more complicated

File: packages/salesforce-adapter/src/fetch.ts (hunk 4 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts:295]: No, folders were fetched only for one of `InFolderMetadataType`. It just happened as a different branch within the method `listMetadataObjects`

File: packages/salesforce-adapter/src/fetch.ts (hunk 5 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts]: If we know this types beforehand not sure why not? 
Also, this is handled within the `metadata_types` module which should be updated every time new types are added (probably in a new API version bump). If they add another type that is stored within Folders, we will update this list accordingly. This makes the code way more safe IMO.

File: packages/salesforce-adapter/src/fetch.ts (hunk 7 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts:235]: I'm passing the fileName in addition to the other things. I use the fileName later in `isInstanceMatchQueryParams` specifically for `InFolderMetadataType | FolderMetadataType`

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 8 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:33]: Renamed to `getFolderPathsByName`. This should return mapping between the FolderName to it's full path.
e.g. 
```
{
  ReportsFolder: 'ReportsFolder',
  NestedFolder: 'ReportsFolder/NestedFolder',
}
```

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 9 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:60]: Renamed to `getAllowedStrings`. This method parses a regex string into string array of valid options

File: packages/salesforce-adapter/src/fetch.ts (hunk 7 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts]: I've simplified my solution to not use the fileName anymore :)

File: packages/salesforce-adapter/src/fetch.ts (hunk 6 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch.ts:166]: [getFolders](https://github.com/salto-io/salto/blob/d825b4d2e671381e4d8074a65269af131c3240a2/packages/salesforce-adapter/src/fetch.ts#L231) used to set the config changes in place and return a list of the fileProperties (a.k.a. elements). Here I merge the results before returning.

File: packages/salesforce-adapter/src/fetch.ts [hunk 10]
```diff
@@ -103,10 +175,10 @@ const getFullName = (obj: FileProperties): string => {
   // in the fullName even when they should have it
   const fullNameWithCorrectedObjectName = obj.fullName.startsWith(namePrefix) ? obj.fullName : `${namePrefix}${obj.fullName}`
   if (obj.type === LAYOUT_TYPE_ID_METADATA_TYPE && obj.namespacePrefix) {
-  // Ensure layout name starts with the namespace prefix if there is one.
-  // needed due to a SF quirk where sometimes layout metadata instances fullNames return as
-  // <namespace>__<objectName>-<layoutName> where it should be
-  // <namespace>__<objectName>-<namespace>__<layoutName>
+    // Ensure layout name starts with the namespace prefix if there is one.
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch.ts:178]: Is this change on purpose?

File: packages/salesforce-adapter/test/adapter.discover.test.ts [hunk 11]
```diff
@@ -36,6 +36,16 @@ import {
 import { FetchElements, MAX_ITEMS_IN_RETRIEVE_REQUEST } from '../src/types'
 import { fetchMetadataInstances } from '../src/fetch'
 import * as fetchModule from '../src/fetch'
+import * as xmlTransformerModule from '../src/transformers/xml_transformer'
+import * as metadataQueryModule from '../src/fetch_profile/metadata_query'
+
+const { makeArray } = collections.array
+
+// const mockedMetadataQueryModule = jest.mocked(metadataQueryModule)
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/test/adapter.discover.test.ts:44]: Should we remove this?

File: packages/salesforce-adapter/src/fetch.ts (hunk 4 above)
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/fetch.ts:298]: How are you sure? This used to happen on all types with an annotation of folderContentType, which is part of the js-force return result.
I think this hard coded list might cause a current regression or a future regression. I'm not sure we can safely define this list of types.

**Review by amitgoldarb — COMMENTED** [review]
Partially reviewed, let's discuss this one f2f. 
I think in this PR we started to ignore the folderType from the js-force result

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 12]
```diff
@@ -105,6 +134,23 @@ export const buildMetadataQuery = (
     ),
 
     isPartialFetch: () => target !== undefined,
+
+    getFolderPathsByName: (type: InFolderMetadataType) => {
+      const folderPaths = include
+        .filter(params => params.metadataType === METADATA_TYPE_TO_FOLDER_TYPE[type])
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:140]: This should be with the annotation of the type

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 13]
```diff
@@ -105,6 +134,23 @@ export const buildMetadataQuery = (
...
+        .flatMap(params => {
+          const { name: nameRegex } = params
+          return isDefined(nameRegex)
+            ? getAllowedStrings(nameRegex)
+            : []
+        })
+      const folderPathsByName = _.keyBy(folderPaths, path => _.last(path.split('/')) ?? path)
+      return _.omitBy(
+        folderPathsByName,
+        // We omit any regex paths from the mapping e.g. .*, FolderName1? etc...
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:150]: Can we explain a bit more why we omit?

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 14]
```diff
@@ -64,9 +89,13 @@ export const buildMetadataQuery = (
...
       ? getDefaultNamespace(instance.metadataType)
       : namespace
-    return regex.isFullRegexMatch(instance.metadataType, metadataType)
-    && regex.isFullRegexMatch(instance.namespace, realNamespace)
-    && regex.isFullRegexMatch(instance.name, name)
+    if (!regex.isFullRegexMatch(instance.metadataType, metadataType)
+      || !regex.isFullRegexMatch(instance.namespace, realNamespace)) {
+      return false
+    }
+    return isFolderMetadataType(instance.metadataType)
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:96]: Use the annotation instead

**Review by amitgoldarb — COMMENTED** [review]
Looks good, discussed f2f.
Fix the few small comments and we should be good here.

File: packages/salesforce-adapter/test/adapter.discover.test.ts (hunk 11 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/adapter.discover.test.ts]: removed

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 12 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts]: v

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 13 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts]: No need to explain anymore, changed `allowedStrings` to `getPaths`. `getPaths` uses the indicative regex `FOLDER_PATH_RE` that explains the valid chars for paths. in SFDX

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 14 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts]: v

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts [hunk 15]
```diff
@@ -42,11 +50,26 @@ const DEFAULT_NAMESPACE_MATCH_ALL_TYPE_LIST = [
...
 const getDefaultNamespace = (metadataType: string): string =>
   (DEFAULT_NAMESPACE_MATCH_ALL_TYPE_LIST.includes(metadataType)
     ? '.*'
     : DEFAULT_NAMESPACE)
 
+const getPaths = (regexString: string): string[] => (
+  regexString
+    .replace(/[()^$]/g, '')
+    .split('|')
+    .filter(path => VALID_FOLDER_PATH_RE.test(path))
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:63]: Is this a must? I think we can remove this no? If it so happens that there exists a folderName without these chars (in some future version), what would be the consequence?

File: packages/salesforce-adapter/src/fetch.ts (hunk 10 above)
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/fetch.ts:163]: ☝️

**Review by amitgoldarb — APPROVED** [review]
LGTM!
Thanks for this PR, great work!
Approved with 2 small comments.

File: packages/salesforce-adapter/src/fetch_profile/metadata_query.ts (hunk 15 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/fetch_profile/metadata_query.ts:63]: discussed f2f
